### DISCUSSION TITLE
EntityRuler and ent_id vs EntityLinker and kb_id

### DISCUSSION THREAD

**Discussion by cwhatley** [explosion/spaCy#8034] [Help: Best practices]
Is there a rationale that one should be aware of behind the fact that entity IDs in the rule based system being attached to spans and tokens in a different place than they are by the statistical/knowledge based system? Are they designed/intended for different things or being stored in different places so the application can choose one over the other if both ruler and linker are used?

Right now, I'm bootstrapping with rules and will probably go to the entity linker later, but it seems like I'll have to go update any references to token.ent_id to be to token.kb_id at that point.

**Comment by svlandeg** [comment]
I think at the time of implementing Named Entity Linking, we felt like we needed a separate field to store actual unique Knowledge Base identifiers, as they are semantically often different to `ent_id`'s, and in some cases you want to be able to have both.

In general I don't think a rule-based approach will allow you to truly bootstrap an entity-linking system, as the latter almost always needs context to do disambiguation. I could imagine that the rules provide the first bit of information, and that you'd need to refine the `ent_id` into a specific `kb_id` for specific mentions in a specific context.

But it's difficult to say without having more specifics on your project ;-) Either way, if you run into specific problems with this, let us know!
